datumaro/components/progress_reporting.py

```diff
@@ -0,0 +1,69 @@
+# Copyright (C) 2022 Intel Corporation
+#
+# SPDX-License-Identifier: MIT
+
+from typing import Iterable, Optional, TypeVar
+import math
+
+T = TypeVar('T')
+
+class ProgressReporter:
+    def get_frequency(self) -> float:
+        """
+        Returns reporting frequency.
+
+        For example, 0.1 would mean every 10%.
+        """
+        raise NotImplementedError
+
+    def start(self, total: int, *, desc: Optional[str] = None):
```

Comment:
IMO, the way that multiple progress bars are handled needs to be reworked. The current approach has two problems:

1. There is no way for the client to gauge how much progress has been made in total, because the caller doesn't know how many times the service (by which I mean the component reporting the progress, such as an extractor) will call `start`. This will be a problem if Datumaro is integrated into a bigger pipeline which only allows reporting progress as a single fraction, such as Accuracy Checker.

2. It's not possible to have multiple concurrently-updating progress bars, because the `report_status` and `finish` functions don't have enough information to determine which bar is being updated.

I propose the following redesign that fixes these problems:

* Remove the `desc` argument from `start` and `iter`.
* Add the following method:

  ```python
  def split(self, subtasks: Sequence[str]) -> Tuple[ProgressReporter, ...]: ...
  ```

  This method tells the reporter that the current task consists of several subtasks (each element of `subtasks` is the description of a subtask) and returns one progress reporter for each subtask.

* Stipulate that the service must call one and only one of the following methods on its reporter:
  * `start` - if it can subdivide its work into N homogenous units. If it calls `start`, then it must use `report_status` to report on the completion of these units.
  * `split` - if it can subdivide its work into heterogenous units (subtasks). If it calls `split`, then no more methods on the original reporter should be called. Instead, the returned reporters must be used by following the rules recursively.

Reply:
Yeah, it makes sense. It still can be hard to solve the first problem with the proposed design. The second problem looks mostly theoretical at this point.

Reply:
> It still can be hard to solve the first problem with the proposed design.

What do you mean? It seems fairly easy to calculate:

* If the service has called `start`, then the progress is the current number of completed units divided by the total number of units (if the service didn't specify the total number, then the progress can only be approximated as 0 before the call to `finish` and 1 after).
* If the service has called `split`, then the progress is the sum of the progresses of the child reporters, divided by the number of child reporters.
* If the service has called neither, the progress is 0.

Reply:
Yes, it is basically correct, but the problem is that if we only have a single pbar, children pbars do one of the following:
- Commit to the single pbar. Depending if we know total or not, there are mixed variants possible. For instance, after starting a new pbar, we can get a new total and resize the pbar, or we can get the situation that we overfill the previous pbar (if we knew total, but now we don't). If we know all totals, we can sum them, of course, but it won't mean that units are equal.
- Split pbar into segments and fill them independently
- Rewrite pbar

I don't consider parallel processing here, but it can be a problem, in theory. Generally, I like the idea, but I'm just saying it's not a silver bullet.

Reply:
I've implemented this idea.

Reply:
Okay, but what about about my proposal about moving the description argument to `split`? I think that makes more sense, because if the service doesn't call `split`, then it doesn't need to supply a description - the client already knows what it's going to be doing. For example, if the client calls a converter, then the client can always label the progress bar "Exporting dataset", because that's what converters do.

On the other hand, descriptions are necessary when you're splitting the task into subtasks, since the client doesn't know what each subtask represents.

Reply:
Probably, it's reasonable to always split the input pbar, so the caller title is not going to be used at all. I don't have enough arguments for or against requiring titles in `split`, but it looks more flexible to just ask the number.